choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java

```
@@ -0,0 +1,87 @@
+    
+    @Override
+    public void startServerObservation(ObserverContext observerContext) {
+        startObservation(observerContext);
+    }
+    @Override
+    public void startClientObservation(ObserverContext observerContext) {
+        startObservation(observerContext);
+    }
+    @Override
+    public void stopServerObservation(ObserverContext observerContext) {
+        if (!observerContext.isStarted()) {
+            // Do not collect metrics if the observation hasn't started
+            return;
+        }
+        stopObservation(observerContext);
+    }
+    @Override
+    public void stopClientObservation(ObserverContext observerContext) {
+        if (!observerContext.isStarted()) {
+            // Do not collect metrics if the observation hasn't started
+            return;
+        }
+        stopObservation(observerContext);
+    }
+    
+    private void startObservation(ObserverContext observerContext) {
+        observerContext.addProperty(PROPERTY_START_TIME, System.nanoTime());
+    }
+    
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = observerContext.getAllTags();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.nanoTime() - startTime;
+            long steps= Math.round(Math.ceil((double)(duration-500)/500));
+            metricRegistry.counter(new MetricId("choreo_steps_total",
+                    "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            handleError("multiple metrics", tags, e);
+        }
+    }
+    
+    private void handleError(String metricName, Set<Tag> tags, RuntimeException e) {
+        // Metric Provider may throw exceptions if there is a mismatch in tags.
+        consoleError.println("error: error collecting metrics for " + metricName + " with tags " + tags +
+                ": " + e.getMessage());
```

Ideally we should log the full stacktrace if the exception is not propagated further